lib/nxdk/log_udp.c

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+
+// SPDX-FileCopyrightText: 2022 Mara (apfel)
+
+#include "log_udp.h"
+
+#include <lwip/arch.h>
+#include <lwip/debug.h>
+#include <lwip/dhcp.h>
+#include <lwip/init.h>
+#include <lwip/netdb.h>
+#include <lwip/netif.h>
+#include <lwip/opt.h>
+#include <lwip/sockets.h>
+#include <lwip/sys.h>
+#include <lwip/tcpip.h>
+#include <lwip/timeouts.h>
+#include <netif/etharp.h>
+#include <pktdrv.h>
+
+#define UDP_USE_DHCP 1
+#define UDP_LOG_PORT 4405
+
+#ifdef NXDK_UDP_USE_INTERNAL_NETIF_IMPL
+struct netif* g_pnetif, nforce_netif;
+
+err_t nforceif_init(struct netif* netif);
+
+static void tcpip_init_done(void* arg)
+{
+	sys_sem_signal((sys_sem_t*)arg);
+}
+
+static void packet_timer(void *arg)
+{
+    LWIP_UNUSED_ARG(arg);
+    Pktdrv_ReceivePackets();
+    sys_timeout(5, packet_timer, NULL);
+}
+#endif
+
+static int udp_socket;
+static struct sockaddr_in udp_sockaddr;
+
+size_t udp_log_function(const char* message, size_t length)
+{
+    return sendto(udp_socket, message, length, 0, (struct sockaddr*)&udp_sockaddr, sizeof(struct sockaddr_in));
+}
+
+uint8_t nxLogUDPRegister(ip4_addr_t* ip, ip4_addr_t* netmask, ip4_addr_t* gateway)
+{
+    // a really disgusting way to handle this, but oh well
+    if ((ip == NULL && (netmask != NULL || gateway != NULL)) || (netmask == NULL && (ip != NULL || gateway != NULL)) || (gateway == NULL && (ip != NULL || netmask != NULL))) return nxLogResult_InvalidArguments;
+
+    uint8_t reg_result = nxLogRegisterFunction(&udp_log_function);
+    nxLogUnregisterFunction(&udp_log_function);
+    switch (reg_result)
+    {
+    case nxLogResult_AlreadyRegistered:
+        return nxLogResult_AlreadyRegistered;
+
+    case nxLogResult_MaximumNumberReached:
+        return nxLogResult_MaximumNumberReached;
+
+    default:
+        break;
+    }
+
+#ifdef NXDK_UDP_USE_INTERNAL_NETIF_IMPL
+    sys_sem_t init_complete;
+    sys_sem_new(&init_complete, 0);
+	tcpip_init(tcpip_init_done, &init_complete);
```

Comment:
I don't think the logging code should do (de-) initialization of the network stack.

Reply:
This was mainly for projects that don't use networking on their own but don't want to use the console log for whatever reason, so that they don't have to initialize it themselves.

Reply:
Maybe a system like the nxdk automount stuff for networking instead?
I agree with thrimbor that most libs should not mess up the network state.